### PR TITLE
work in progress

### PR DESCRIPTION
**Pull request by antoniocarlosortiz**

### CODE REVIEW

**Comment by antoniocarlosortiz** [comment]
@benigeri for your review.

File: app/api/companies/forms.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+from flask_wtf import Form
+from wtforms import StringField
+from wtforms.validators import Required, Length, Regexp, URL
+
+class CreateCompanyForm(Form):
+	name = StringField('name', validators=[
+			Required(message="Company name is required to create an entry."),
```
Comment (benigeri) [inline app/api/companies/forms.py]: Would change message to "Company name is required."

File: app/api/companies/forms.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+			Length(max=60, message="Company name can't be longer than 60 characters."),
+			Regexp("^[A-Za-z0-9_]+$",
```
Comment (benigeri) [inline app/api/companies/forms.py]: I don't think we need this validation. Some companies might have weird symbols in their names, and spaces....

File: app/api/companies/forms.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+				   message="Car name can only use letters, numbers, and '_'")])
+
+	url = StringField('url', validators=[
+			Required(message="url is required to create an entry."),
```
Comment (benigeri) [inline app/api/companies/forms.py]: Error message should be same format as "Company name is required."

File: default_config.ini [hunk 4]
```diff
@@ -15,8 +15,8 @@ version = 0.0
 [db]
 username =
 password =
-name = flask_api_template
-url = postgresql+psycopg2://antonio:stangg123@localhost/flask_api_template
```
Comment (benigeri) [inline default_config.ini]: default config should not include your information, it should be generic stuff. You should put this in your local_config.ini

File: backend/models.py [hunk 5]
```diff
@@ -6,9 +6,16 @@
...
 
-class Car(Base):
-    __tablename__ = 'car'
-    car_id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=True)
-    name = sa.Column(sa.String(length=60), nullable=False, unique=True)
-    created_at = sa.Column(sa.DateTime, nullable=False,
+class EmployeeCountEntry(Base):
+    __tablename__ = 'employee_count_entry'
+    entry_id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=True)
+    company_name = sa.Column(sa.String(length=60), nullable=False, unique=True)
```
Comment (benigeri) [inline backend/models.py]: Should be company_id

Also you need to manually create an index to make searching efficient: http://docs.sqlalchemy.org/en/latest/core/constraints.html#indexes

File: backend/models.py [hunk 6]
```diff
@@ -6,9 +6,16 @@
...
 
 
-class Car(Base):
-    __tablename__ = 'car'
-    car_id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=True)
-    name = sa.Column(sa.String(length=60), nullable=False, unique=True)
-    created_at = sa.Column(sa.DateTime, nullable=False,
+class EmployeeCountEntry(Base):
+    __tablename__ = 'employee_count_entry'
+    entry_id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=True)
```
Comment (benigeri) [inline backend/models.py]: should be employee count id

File: backend/employee_count_entry.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+from backend import db, models
+from shared_libs import exceptions
+
+def create_entry(company_name):
```
Comment (benigeri) [inline backend/employee_count_entry.py]: Should take in company_id, not company name. Also needs an employee count number, no?

File: backend/employee_count_entry.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    name = company_name.lower()
+
+    with db.scoped_session() as session:
+        new_entry = models.EmployeeCountEntry(entry_id=entry)
```
Comment (benigeri) [inline backend/employee_count_entry.py]: where are you getting entry from? This code doesn't run does it?

File: backend/employee_count_entry.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        session.add(new_entry)
+        session.commit()
+        return new_company_entry.new_entry
+
+
+def get_entry(entry_id):
+    with db.scoped_session() as session:
+        entry = session.query(models.EmployeeCountEntry).\
+                filter_by(entry_id=entry_id).first()
```
Comment (benigeri) [inline backend/employee_count_entry.py]: entry_id is too generic, need a better name.

File: backend/alembic/versions/d27619e9be3_create_user_table.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
+"""create user table
+
+Revision ID: d27619e9be3
```
Comment (benigeri) [inline backend/alembic/versions/d27619e9be3_create_user_table.py]: why is this called crete user table? you can name the migration using the alembic command. Let me know if you need help with that.

File: backend/alembic/versions/d27619e9be3_create_user_table.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    sa.PrimaryKeyConstraint('entry_id'),
+    sa.UniqueConstraint('company_name')
+    )
+    op.drop_table('car')
+    ### end Alembic commands ###
+
+
+def downgrade():
+    ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('car',
```
Comment (benigeri) [inline backend/alembic/versions/d27619e9be3_create_user_table.py]: You should't have any of the car stuff in here. You should start from a blank database.

File: backend/alembic/versions/4265265b454b_create_user_table.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
+"""create user table
```
Comment (benigeri) [inline backend/alembic/versions/4265265b454b_create_user_table.py]: This shouldn't be in here.

File: backend/alembic/versions/3a6952426f81_create_user_table.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
+"""create user table
```
Comment (benigeri) [inline backend/alembic/versions/3a6952426f81_create_user_table.py]: why is this called create user table? 

And you should combine migrations into one if you mess them up / forget something: 

1. downgrade using alembic (alembic downgrade -1 or something)
2. then delete the migration
3. create a new migration with all of the changes you like

File: backend/alembic/versions/28399fc0600a_create_user_table.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
+"""create user table
```
Comment (benigeri) [inline backend/alembic/versions/28399fc0600a_create_user_table.py]: why is this here?

File: backend/alembic/versions/12d627ec08a_create_user_table.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
+"""create user table
```
Comment (benigeri) [inline backend/alembic/versions/12d627ec08a_create_user_table.py]: Shouldn't be called create_user_table....

File: app/api/employee_count_entry/forms.py [hunk 16]
```diff
@@ -0,0 +1,20 @@
+from flask_wtf import Form
+from wtforms import StringField, IntegerField, DateField
+from wtforms.validators import Required, Length, Regexp
+
+
+class CreateEntryForm(Form):
+    company_name = StringField('name', validators=[
+    				Required(message='company name is required to create an entry.'),
```
Comment (benigeri) [inline app/api/employee_count_entry/forms.py]: see comments on model and update. Also see other comments for form error message style.

File: app/api/employee_count_entry/__init__.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
+from flask import Blueprint
+from app import json_response
+from app.api.employee_count_entry import forms
+
+from backend import employee_count_entry as entries_controller
+
+entries = Blueprint('entries', __name__)
+
+# SQLAlchemy Objects are not json serializable
```
Comment (benigeri) [inline app/api/employee_count_entry/__init__.py]: You don't need to to include this comment everytime, it was just for you.

File: app/api/employee_count_entry/__init__.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    entries_json = [entry_to_json(entry) for entry in entry_list]
+    return json_response.success({'entries': entries_json})
+
+
+@entries.route('/<int:entry_id>', methods=['GET'])
+def get_entry(entry_id):
+    entry = entries_controller.get_entry(entry_id)
+    return json_response.success({'entry': entry_to_json(entry)})
+
+@entries.route('/<int:entry_id>', methods=['DELETE'])
```
Comment (benigeri) [inline app/api/employee_count_entry/__init__.py]: I don't think we need an endpoint to delete entries. You should remove that function in the controller too.

**Comment by benigeri** [comment]
@antoniocarlosortiz a few comments to fix :) Make sure to fix them on this branch/pr. Do not start working or merge this until everything is resolved, and it passes the circle build. 

You also need to write some basic test cases for all of the endpoints. 

You did not run tox. You have tons of style errors, and I think that some of your code doesn't even run without crashing. Did you test this at all? In the future, you code should pass tox and be tested before I review it. 

Good start! Lots of work and learning ahead of us though.

File: app/api/companies/forms.py (hunk 1 above)
Comment (antoniocarlosortiz) [inline app/api/companies/forms.py]: Done.

File: app/api/companies/forms.py (hunk 2 above)
Comment (antoniocarlosortiz) [inline app/api/companies/forms.py]: Youre right. I was excited to use regex. Done.

File: app/api/companies/forms.py (hunk 3 above)
Comment (antoniocarlosortiz) [inline app/api/companies/forms.py]: Done.

File: default_config.ini (hunk 4 above)
Comment (antoniocarlosortiz) [inline default_config.ini]: Is it alright if you could help me override the section that calls this file so i can override it to look for a local version instead or do you think it would be better if I just have git untrack the default_config.ini and track instead a placeholder ini file that users could copy to the 'default_config.ini' on their local?

File: backend/models.py (hunk 5 above)
Comment (antoniocarlosortiz) [inline backend/models.py]: Thanks. Ill read this. I will also delete EmployeeCountEntry. These are code that ive written before the sync with you.

File: backend/models.py (hunk 6 above)
Comment (antoniocarlosortiz) [inline backend/models.py]: Ill delete EmployeeCountEntry Model and create a new one that is more similar to the project spec

File: backend/employee_count_entry.py (hunk 7 above)
Comment (antoniocarlosortiz) [inline backend/employee_count_entry.py]: Ill delete this file and create a new one that is fashioned with the project spec.

File: backend/alembic/versions/d27619e9be3_create_user_table.py (hunk 10 above)
Comment (antoniocarlosortiz) [inline backend/alembic/versions/d27619e9be3_create_user_table.py]: Ohh okay you caught me mindlessly copying the command on the README.md

**Comment by antoniocarlosortiz** [comment]
@benigeri I didnt test the parts that Im planning on deleting cos those were the sections that I've written before my sync with you. Ill do better next time sorry sorry.

**Comment by antoniocarlosortiz** [comment]
@benigeri hey Paul. For your review.

File: default_config.ini (hunk 4 above)
Comment (benigeri) [inline default_config.ini]: Check out https://github.com/SiliconValleyInsight/nodes-and-edges/blob/master/shared_libs/config.py

Everything thing in local_config.ini should override default_config.ini 

So default_config.ini should contain placeholders for all the values, and local_config.ini should have the blanks filled in. For example in default_config you'd have: 

```
[fb]
app_name = TheAppName
api_key = 
```
And in local_config.ini you'd have
```
[fb]
api_key = 123123
```

Then the config file will get the fb.app_name from default, and will get fb.api_key from local. 

If you wanted to override the app name too, you could do: 
```
[fb]
app_name = TheOverridenAppName
api_key = 123123123
```

File: backend/alembic/versions/d27619e9be3_create_user_table.py (hunk 10 above)
Comment (benigeri) [inline backend/alembic/versions/d27619e9be3_create_user_table.py]: yup, there's an argument for the migration name

File: app/api/companies/__init__.py [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+
+companies = Blueprint('companies', __name__)
+
+
+def company_to_json(company):
+    return {'company_id': company.company_id,
+            'name': company.name,
+            'url': company.url
+            }
```
Comment (benigeri) [inline app/api/companies/__init__.py]: Closing } should be on same line as company.url: 

```
+    return {'company_id': company.company_id,
+            'name': company.name,
+            'url': company.url}
```

File: app/api/companies/forms.py [hunk 20]
```diff
@@ -0,0 +1,15 @@
+from flask_wtf import Form
+from wtforms import StringField
+from wtforms.validators import Required, Length, URL
+
+
+class CreateCompanyForm(Form):
+    name = StringField('name', validators=[
+            Required(message="Company name is required."),
+            Length(max=60, message="Company name can't be longer than 60 characters."),
+            ])
```
Comment (benigeri) [inline app/api/companies/forms.py]: see above

File: app/api/companies/forms.py [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+
+    url = StringField('url', validators=[
+            Required(message="URL is required."),
+            URL(message="invalid URL")
+            ])
```
Comment (benigeri) [inline app/api/companies/forms.py]: see above

File: backend/companies.py [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+        session.add(new_company)
+        session.commit()
+        return new_company.company_id
+
+
+def get_company(company_id):
+    with db.scoped_session() as session:
+        company = session.query(models.Company).\
+                    filter_by(company_id=company_id).first()
+        session.expunge(company)
```
Comment (benigeri) [inline backend/companies.py]: you should read up on what expunge does, it's useful to know
http://docs.sqlalchemy.org/en/improve_toc/orm/session_state_management.html

File: backend/companies.py [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def delete_company(company_id):
+    with db.scoped_session() as session:
+        company = session.query(models.Company).\
+                    filter_by(company_id=company_id).first()
+
+        if company is None:
+            raise exceptions.AppError(
+                "The company you tried to delete does not exist")
```
Comment (benigeri) [inline backend/companies.py]: wherez de full stop?

File: backend/models.py [hunk 24]
```diff
@@ -1,14 +1,12 @@
...
-class Car(Base):
-    __tablename__ = 'car'
-    car_id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=True)
+class Company(Base):
+    __tablename__ = 'company'
+    company_id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=True)
     name = sa.Column(sa.String(length=60), nullable=False, unique=True)
-    created_at = sa.Column(sa.DateTime, nullable=False,
-                           default=datetime.datetime.utcnow)
+    url = sa.Column(sa.String, nullable=False)
```
Comment (benigeri) [inline backend/models.py]: We should still have the timestamp.

File: default_config.ini (hunk 4 above)
Comment (benigeri) [inline default_config.ini]: PS this is causing the circleci build to crash.

File: testing_config.ini [hunk 25]
```diff
@@ -4,7 +4,7 @@ testing = true
 
 [db]
 name = flask_api_template_test
-url = postgresql+psycopg2://localhost/flask_api_template_test
+url = postgresql+psycopg2://antonio:stangg123@localhost/flask_api_template_test
```
Comment (benigeri) [inline testing_config.ini]: Oh no.. this is the problem. Let's sync tonight to see how we can fix this.

File: tests/api/test_companies.py [hunk 26]
```diff
@@ -0,0 +1,114 @@
+import unittest
+
+from tests.api import APITestCase, responses
+from backend import companies
+
+
+def create_default_company():
+    name = 'dogindustries'
+    url = 'https://dogs.com'
```
Comment (benigeri) [inline tests/api/test_companies.py]: url should be a linked in company url: https://www.linkedin.com/company/black-dog-industries-llc

Comment (benigeri) [inline tests/api/test_companies.py]: this is a bad example though because there's no employees

**Comment by benigeri** [comment]
cool beans, getting close @antoniocarlosortiz 

btw, the average PR review time goes down exponentially. They get easier and easier as you learn :)

File: app/api/companies/__init__.py (hunk 19 above)
Comment (antoniocarlosortiz) [inline app/api/companies/__init__.py]: i thought it was the style pep8 wanted hahah alright ill change it :D

File: backend/companies.py (hunk 23 above)
Comment (antoniocarlosortiz) [inline backend/companies.py]: Thanks. Ill add a return command.

File: testing_config.ini (hunk 25 above)
Comment (stefsy) [inline testing_config.ini]: @antoniocarlosortiz please explain the fix to me later! I'm curious too

File: default_config.ini (hunk 4 above)
Comment (antoniocarlosortiz) [inline default_config.ini]: Thanks for the explanation! Hey can ask what about for the testing_config? Does it also have its own version of local_config?

**Comment by antoniocarlosortiz** [comment]
@benigeri btw i didnt create any endpoint for the employee count model since its not specified in the spec. Would that be alright?

File: default_config.ini (hunk 4 above)
Comment (benigeri) [inline default_config.ini]: any update on getting your db to work with the system default username + no word?

File: app/api/companies/__init__.py [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+
+def company_to_json(company):
+    employee_count_list = companies_controller.\
+                            get_employee_counts(company.company_id)
+
+    if employee_count_list:
+        employee_counts_json = [employee_count_to_json(employee_count)
+                                for employee_count in employee_count_list]
+    else:
+        employee_counts_json = None
```
Comment (benigeri) [inline app/api/companies/__init__.py]: should return [] instead of none

File: app/api/companies/__init__.py [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def employee_count_to_json(employee_count):
+    return {'employee_count_id': employee_count.employee_count_id,
+            'timestamp': str(employee_count.timestamp),
+            'number': employee_count.number,
+            'company_id': employee_count.company_id}
+
+
+def company_to_json(company):
```
Comment (benigeri) [inline app/api/companies/__init__.py]: this function should just deal with formatting things to json. It's better to pass in the data than to get it within this function.

File: app/api/companies/__init__.py [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+        return json_response.success({'company_id': company_id})
+
+    else:
+        return json_response.form_errors(create_company_form)
+
+
+@companies.route('', methods=['GET'])
+def get_companies():
+    company_list = companies_controller.get_companies()
+    companies_json = [company_to_json(company) for company in company_list]
```
Comment (benigeri) [inline app/api/companies/__init__.py]: This is very inefficient because we have to make a query for each company. For now, this route should only return data from the company model. We can worry about getting some employee count data later.

File: app/api/companies/forms.py [hunk 30]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class CreateCompanyForm(Form):
+    name = StringField('name', validators=[
+            Required(message="Company name is required."),
+            Length(max=60, message="Company name can't be longer than 60 characters.")])
+
+    url = StringField('url', validators=[
+            Required(message="URL is required."),
+            URL(message="invalid URL")])
```
Comment (benigeri) [inline app/api/companies/forms.py]: Dude write nicer error messages! 

"The company URL is invalid."

File: backend/alembic/versions/213e4ab159fa_create_employeecount_table.py [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def upgrade():
+    ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('employee_count',
+    sa.Column('employee_count_id', sa.BigInteger(), nullable=False),
+    sa.Column('timestamp', sa.Date(), nullable=False),
+    sa.Column('number', sa.Integer(), nullable=False),
+    sa.Column('company_id', sa.BigInteger(), nullable=True),
+    sa.ForeignKeyConstraint(['company_id'], ['company.company_id'], ),
```
Comment (benigeri) [inline backend/alembic/versions/213e4ab159fa_create_employeecount_table.py]: You should get rid of this. This is makes database sharing very difficult. It's also not necessary.

File: app/api/companies/__init__.py (hunk 28 above)
Comment (antoniocarlosortiz) [inline app/api/companies/__init__.py]: Ok thanks I'll create a new function then pass the data to company_to_json. What do you think of my naming conventions btw?

File: app/api/companies/__init__.py (hunk 27 above)
Comment (antoniocarlosortiz) [inline app/api/companies/__init__.py]: Okay thanks. I actually spent some time what would be most appropriate. Thanks for pointing this out.

File: app/api/companies/__init__.py (hunk 29 above)
Comment (antoniocarlosortiz) [inline app/api/companies/__init__.py]: Oh ok got it. Yes it does create a query for every company. Thanks.

File: app/api/companies/forms.py (hunk 30 above)
Comment (antoniocarlosortiz) [inline app/api/companies/forms.py]: Haha alright my bad. It sounded right in my mind like those robots you see on tv that encounters something unexpected "ERROR ERROR" haha

File: backend/alembic/versions/213e4ab159fa_create_employeecount_table.py (hunk 31 above)
Comment (antoniocarlosortiz) [inline backend/alembic/versions/213e4ab159fa_create_employeecount_table.py]: This is the file created by the migration command. I'm not sure if I understood it correctly but should I remove the EmployeeCount Table I created and place all the columns in it instead on the Companies Table. It does make database sharing difficult.

File: default_config.ini (hunk 4 above)
Comment (antoniocarlosortiz) [inline default_config.ini]: Yep! I edited the config files pg_hba.conf to not ask for password and I changed the owner of the test databases to Postgres so I don't have to write the name in the test_config url.

File: backend/companies.py (hunk 22 above)
Comment (antoniocarlosortiz) [inline backend/companies.py]: I read about this. Basically expunge removes the object from the session but that object is still in the database. Is it alright if I why this would be useful in that function?

File: app/api/companies/__init__.py (hunk 28 above)
Comment (benigeri) [inline app/api/companies/__init__.py]: cool. What naming conventions?

File: backend/models.py [hunk 32]
```diff
@@ -1,14 +1,24 @@
...
 
-class Car(Base):
-    __tablename__ = 'car'
-    car_id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=True)
+class Company(Base):
+    __tablename__ = 'company'
+    company_id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=True)
     name = sa.Column(sa.String(length=60), nullable=False, unique=True)
+    url = sa.Column(sa.String, nullable=False)
+    employee_counts = relationship('EmployeeCount', backref='companies', lazy='dynamic')
```
Comment (benigeri) [inline backend/models.py]: ### we shouldn't have these kind of relationships in here. We can just query them directly, so no need to add this kind of overhead.

File: backend/alembic/versions/213e4ab159fa_create_employeecount_table.py (hunk 31 above)
Comment (benigeri) [inline backend/alembic/versions/213e4ab159fa_create_employeecount_table.py]: ah, see comments that start with ^^^

File: backend/companies.py (hunk 22 above)
Comment (benigeri) [inline backend/companies.py]: let's talk about this on skype

File: backend/models.py [hunk 33]
```diff
@@ -1,14 +1,24 @@
...
     created_at = sa.Column(sa.DateTime, nullable=False,
                            default=datetime.datetime.utcnow)
+
+
+class EmployeeCount(Base):
+    __tablename__ = 'employee_count'
+    employee_count_id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=True)
+    timestamp = sa.Column(sa.Date, nullable=False)
+    number = sa.Column(sa.Integer, nullable=False)
+    company_id = sa.Column(sa.BigInteger, sa.ForeignKey('company.company_id'))
```
Comment (benigeri) [inline backend/models.py]: ^^^ ditch the foreign key here too :)

File: default_config.ini (hunk 4 above)
Comment (benigeri) [inline default_config.ini]: sweet!

File: tests/api/test_companies.py [hunk 34]
```diff
@@ -0,0 +1,174 @@
+import unittest
```
Comment (benigeri) [inline tests/api/test_companies.py]: let's talk about testing on skype

File: tox.ini [hunk 35]
```diff
@@ -6,13 +6,13 @@ envlist = pylint,pep8,tests
 deps =
   -r{toxinidir}/requirements.txt
   pylint==1.1.0
-commands = pylint --rcfile=pylintrc -r n app backend shared_libs tests
+commands = pylint --rcfile=pylintrc -r n app backend shared_libs tests scrape
```
Comment (benigeri) [inline tox.ini]: scrape stuff shouldn't be in this PR

File: tests/api/test_companies.py [hunk 36]
```diff
@@ -0,0 +1,174 @@
...
+        company_id = create_default_employee_count(company.company_id)
+
+        r = self._get_company(company.company_id)
+        data = responses.expect_success(r)
+        company = data.get('company')
+        employee_counts = company.get('employee_counts')
+
+        assert len(employee_counts) == 2
+
+    def test_create_employee_count(self):
```
Comment (benigeri) [inline tests/api/test_companies.py]: you need to clean up the tests as we discussed on Skype. Let me know if you want a reminder in case you didn't have time to write everything :)

**Comment by antoniocarlosortiz** [comment]
@benigeri hey paul for your review. I also keep getting an error in lint on my test_employee_counts.py stating redefine the outer name employee_counts at line 3. I tried googling for the answer and it says that im redeclaring that global variable but I don't see in the code how I'm doing that.

**Comment by benigeri** [comment]
@antoniocarlosortiz did you ever fix the above?

File: tests/api/test_companies.py [hunk 37]
```diff
@@ -0,0 +1,115 @@
...
+    def test_delete_companies(self):
+        # Fail without company_id
+        r = self._delete_company(None)
+        responses.expect_error(r, code=404)
+
+        # Fail deleting a non existant company_id
+        r = self._delete_company(12)
+        responses.expect_error(r)
+
+        company_tuple = create_default_company()
```
Comment (benigeri) [inline tests/api/test_companies.py]: it's more clear to do:

```
a, b = create_default_company()
```

obviously a/b should be appropriate names

File: tests/api/test_companies.py [hunk 38]
```diff
@@ -0,0 +1,115 @@
...
+        # Fail with no name and url
+        r = self._create_company(name=None, url=None)
+        responses.expect_error(r)
+
+        # Fail with no url
+        r = self._create_company(name='cat', url=None)
+        responses.expect_error(r)
+
+        name = 'tomato_org'
+        url = 'http://tomato.com'
```
Comment (benigeri) [inline tests/api/test_companies.py]: should be linkedin url. Otherwise people that read this might get a false impression.

File: tests/api/test_employee_counts.py [hunk 39]
```diff
@@ -0,0 +1,52 @@
+import unittest
+from tests.api import APITestCase, responses
+from backend import companies, employee_counts
+
+
+def create_default_company():
+    name = 'dogindustries'
+    url = 'https://dogs.com'
```
Comment (benigeri) [inline tests/api/test_employee_counts.py]: same here

File: tests/test_employee_counts.py [hunk 40]
```diff
@@ -0,0 +1,58 @@
+import unittest
```
Comment (benigeri) [inline tests/test_employee_counts.py]: I don't think we should have these test. Just having the api test cases is sufficient, and this is more code to maintain. I suggest we remove this file.

File: tests/api/test_employee_counts.py [hunk 41]
```diff
@@ -0,0 +1,52 @@
...
+        company_id = company_tuple[0]
+        company = companies.get_company(company_id)
+
+        company_id = create_default_employee_count(company.company_id)
+        company_id = create_default_employee_count(company.company_id)
+
+        r = self._get_company(company.company_id)
+        data = responses.expect_success(r)
+        company = data.get('company')
+        employee_counts = company.get('employee_counts')
```
Comment (benigeri) [inline tests/api/test_employee_counts.py]: you've already declared employee_counts in line 3. pick a diff name.

**Comment by benigeri** [comment]
@antoniocarlosortiz some small changes. Make sure the tests pass and we can merge it tomorrow!

File: tests/test_employee_counts.py (hunk 40 above)
Comment (antoniocarlosortiz) [inline tests/test_employee_counts.py]: thanks. I'll just put it in gitignore

File: tests/api/test_employee_counts.py (hunk 41 above)
Comment (antoniocarlosortiz) [inline tests/api/test_employee_counts.py]: thanks! I must have been really for not being able to find this.

File: tests/api/test_companies.py (hunk 38 above)
Comment (antoniocarlosortiz) [inline tests/api/test_companies.py]: got it. Already changed

File: tests/api/test_companies.py (hunk 37 above)
Comment (antoniocarlosortiz) [inline tests/api/test_companies.py]: got it.

**Comment by benigeri** [comment]
looks good!
